File: app/src/main/java/configgen/editorserver/TableCreateService.java
package configgen.editorserver;

import com.alibaba.fastjson2.JSONArray;
import com.alibaba.fastjson2.JSONObject;
import configgen.schema.*;
import configgen.schema.cfg.CfgReader;
import configgen.schema.cfg.CfgSyntaxException;
import configgen.schema.cfg.CfgWriter;
import configgen.tool.SchemaToCsvHeader;
import configgen.util.CSVUtil;
import configgen.util.CachedFiles;
import configgen.util.Logger;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

/**
 * 建表服务：接收 JSON 描述的表/结构体/枚举定义，合并到现有 CfgSchema 后写回 config.cfg，
 * 并可选地为 table 类型创建空 CSV 数据文件（仅含表头行）。
 */
public class TableCreateService {

    public record CreateResult(boolean ok, List<String> errors) {
    }

    private static final String IDENTIFIER_PATTERN = "[a-zA-Z_][a-zA-Z0-9_]*";

    /**
     * 创建新表/结构体/枚举并写回 config.cfg。
     *
     * @param dataDir      数据目录（config.cfg 所在目录）
     * @param existingText 当前 config.cfg 文本（用于增量合并）
     * @param request      JSON 请求体，格式：
     *                     { "type": "table"|"struct"|"enum", "name": "mytable",
     *                       "fields": [{ "name": "id", "type": "int", "comment": "..." }],
     *                       "primaryKey": ["id"], "withDataFile": true,
     *                       "enumValues": [{ "name": "A", "comment": "..." }] }
     */
    public static CreateResult createTable(Path dataDir, String existingText, JSONObject request) {
        List<String> errors = new ArrayList<>();

        // 1. 解析现有 config.cfg
        CfgSchema schema;
        try {
            if (existingText != null && !existingText.isBlank()) {
                schema = CfgReader.parse(existingText);
            } else {
                schema = CfgSchema.of();
            }
        } catch (CfgSyntaxException e) {
            errors.add("Failed to parse existing config.cfg: " + e.getMessage());
            return new CreateResult(false, errors);
        } catch (Exception e) {
            errors.add("Failed to parse existing config.cfg: " + e.getMessage());
            return new CreateResult(false, errors);
        }

        // 2. 根据 type 构建新元素
        String type = request.getString("type");
        String name = request.getString("name");

        if (name == null || name.isBlank()) {
            errors.add("Name is required");
            return new CreateResult(false, errors);
        }
        if (!name.matches(IDENTIFIER_PATTERN)) {
            errors.add("Name must be a valid identifier: " + name);
            return new CreateResult(false, errors);
        }

        // 检查名字是否与现有元素冲突
        for (Nameable item : schema.items()) {
            if (item.name().equalsIgnoreCase(name)) {
                errors.add("Name already exists: " + name);
                return new CreateResult(false, errors);
            }
        }

        try {
            Nameable newElement = switch (type) {
                case "table" -> buildTableSchema(request, name);
                case "struct" -> buildStructSchema(request, name);
                case "enum" -> buildEnumTableSchema(request, name);
                default -> {
                    errors.add("Unknown type: " + type + ", must be 'table', 'struct', or 'enum'");
                    yield null;
                }
            };

            if (newElement == null) {
                return new CreateResult(false, errors);
            }

            schema.add(newElement);

            // 3. 校验合并后的 schema
            CfgSchemaErrs errs = schema.resolve();
            try {
                errs.checkErrors("createTable");
            } catch (CfgSchemaException e) {
                for (CfgSchemaErrs.Err err : e.getErrs().errs()) {
                    errors.add(err.msg());
                }
                return new CreateResult(false, errors);
            }

            // 4. 可选：先创建空数据文件（失败则中止，避免 config.cfg 已写入但 CSV 缺失的半成品）
            if ("table".equals(type) && request.getBooleanValue("withDataFile", false)) {
                TableSchema table = (TableSchema) newElement;
                if (!table.meta().hasEnumValues()) {
                    createEmptyCsv(dataDir, table);
                }
            }

            // 5. 写回 config.cfg
            String cfgText = CfgWriter.stringify(schema);
            Path cfgPath = dataDir.resolve("config.cfg");
            CachedFiles.writeFile(cfgPath, cfgText.getBytes(StandardCharsets.UTF_8));
            Logger.log("Table created and config.cfg written: " + name);

            return new CreateResult(true, List.of());

        } catch (CfgSchemaException e) {
            for (CfgSchemaErrs.Err err : e.getErrs().errs()) {
                errors.add(err.msg());
            }
            return new CreateResult(false, errors);
        } catch (Exception e) {
            errors.add(e.getMessage());
            return new CreateResult(false, errors);
        }
    }

    /**
     * 为已有表创建空数据文件（CSV 仅含表头）。
     *
     * @param dataDir  数据目录
     * @param tableName 表名
     */
    public static CreateResult createDataFile(Path dataDir, String tableName) {
        List<String> errors = new ArrayList<>();

        // 读取并解析现有 schema
        Path cfgPath = dataDir.resolve("config.cfg");
        String cfgText;
        try {
            cfgText = Files.readString(cfgPath, StandardCharsets.UTF_8);
        } catch (IOException e) {
            errors.add("Failed to read config.cfg: " + e.getMessage());
            return new CreateResult(false, errors);
        }

        CfgSchema schema;
        try {
            schema = CfgReader.parse(cfgText);
            schema.resolve().checkErrors("createDataFile");
        } catch (Exception e) {
            errors.add("Failed to parse/resolve schema: " + e.getMessage());
            return new CreateResult(false, errors);
        }

        // 找到目标表
        TableSchema targetTable = null;
        for (Nameable item : schema.items()) {
            if (item instanceof TableSchema ts && ts.name().equals(tableName)) {
                targetTable = ts;
                break;
            }
        }

        if (targetTable == null) {
            errors.add("Table not found: " + tableName);
            return new CreateResult(false, errors);
        }

        if (targetTable.isJson()) {
            errors.add("Table uses JSON data source, cannot create CSV file: " + tableName);
            return new CreateResult(false, errors);
        }

        if (targetTable.meta().hasEnumValues()) {
            errors.add("Enum table does not need a data file: " + tableName);
            return new CreateResult(false, errors);
        }

        // 检查文件是否已存在
        Path csvPath = dataDir.resolve(tableName + ".csv");
        if (Files.exists(csvPath)) {
            errors.add("Data file already exists: " + csvPath.getFileName());
            return new CreateResult(false, errors);
        }

        try {
            createEmptyCsv(dataDir, targetTable);
            Logger.log("Data file created: " + csvPath);
            return new CreateResult(true, List.of());
        } catch (IOException e) {
            errors.add("Failed to create data file: " + e.getMessage());
            return new CreateResult(false, errors);
        }
    }

    // ---- 内部方法 ----

    private static TableSchema buildTableSchema(JSONObject request, String name) {
        // table 名必须全小写
        if (!name.equals(name.toLowerCase())) {
            CfgSchemaErrs errs = CfgSchemaErrs.of();
            errs.addErr(new CfgSchemaErrs.TableNameNotLowerCase(name));
            throw new CfgSchemaException(errs);
        }

        JSONArray fieldsArr = request.getJSONArray("fields");
        if (fieldsArr == null || fieldsArr.isEmpty()) {
            throw new IllegalArgumentException("Table must have at least one field");
        }

        List<FieldSchema> fields = new ArrayList<>();
        for (int i = 0; i < fieldsArr.size(); i++) {
            JSONObject f = fieldsArr.getJSONObject(i);
            fields.add(buildFieldSchema(f));
        }

        // 主键
        JSONArray pkArr = request.getJSONArray("primaryKey");
        KeySchema primaryKey;
        if (pkArr != null && !pkArr.isEmpty()) {
            List<String> pkFields = new ArrayList<>();
            for (int i = 0; i < pkArr.size(); i++) {
                pkFields.add(pkArr.getString(i));
            }
            primaryKey = new KeySchema(pkFields);
        } else {
            primaryKey = new KeySchema(List.of(fields.get(0).name()));
        }

        return new TableSchema(
                name.toLowerCase(),
                primaryKey,
                EntryType.ENo.NO,
                false,
                Metadata.of(),
                fields,
                List.of(),
                List.of()
        );
    }

    private static StructSchema buildStructSchema(JSONObject request, String name) {
        JSONArray fieldsArr = request.getJSONArray("fields");
        if (fieldsArr == null || fieldsArr.isEmpty()) {
            throw new IllegalArgumentException("Struct must have at least one field");
        }

        List<FieldSchema> fields = new ArrayList<>();
        for (int i = 0; i < fieldsArr.size(); i++) {
            JSONObject f = fieldsArr.getJSONObject(i);
            fields.add(buildFieldSchema(f));
        }

        return new StructSchema(
                name,
                FieldFormat.AutoOrPack.AUTO,
                Metadata.of(),
                fields,
                List.of()
        );
    }

    private static TableSchema buildEnumTableSchema(JSONObject request, String name) {
        // enum 是一种 TableSchema，通过 Metadata 标记
        if (!name.equals(name.toLowerCase())) {
            CfgSchemaErrs errs = CfgSchemaErrs.of();
            errs.addErr(new CfgSchemaErrs.TableNameNotLowerCase(name));
            throw new CfgSchemaException(errs);
        }

        JSONArray enumValuesArr = request.getJSONArray("enumValues");
        if (enumValuesArr == null || enumValuesArr.isEmpty()) {
            throw new IllegalArgumentException("Enum must have at least one value");
        }

        List<Metadata.EnumValueAssigned> values = new ArrayList<>();
        for (int i = 0; i < enumValuesArr.size(); i++) {
            JSONObject v = enumValuesArr.getJSONObject(i);
            String vName = v.getString("name");
            String vComment = v.getString("comment", "");
            values.add(new Metadata.EnumValueAssigned(vName, vComment, i));
        }

        Metadata meta = Metadata.of();
        meta.putEnumValues(new Metadata.MetaEnumValues.OfAssigned(values));

        // enum 的字段就是 name 列
        FieldSchema nameField = new FieldSchema("name", FieldType.Primitive.STRING,
                FieldFormat.AutoOrPack.AUTO, Metadata.of());

        return new TableSchema(
                name.toLowerCase(),
                new KeySchema(List.of("name")),
                new EntryType.EEnum("name"),
                false,
                meta,
                List.of(nameField),
                List.of(),
                List.of()
        );
    }

    private static FieldSchema buildFieldSchema(JSONObject f) {
        String fieldName = f.getString("name");
        if (fieldName == null || fieldName.isBlank()) {
            throw new IllegalArgumentException("Field name is required");
        }
        if (!fieldName.matches(IDENTIFIER_PATTERN)) {
            throw new IllegalArgumentException("Invalid field name: " + fieldName);
        }

        String fieldType = f.getString("type");
        if (fieldType == null) {
            fieldType = "int";
        }

        FieldType ft = parseFieldType(fieldType);
        Metadata meta = Metadata.of();

        String comment = f.getString("comment");
        if (comment != null && !comment.isBlank()) {
            meta.putComment(new CommentData("", comment, null));
        }

        return new FieldSchema(fieldName, ft, FieldFormat.AutoOrPack.AUTO, meta);
    }

    private static FieldType parseFieldType(String type) {
        return switch (type.toLowerCase()) {
            case "bool" -> FieldType.Primitive.BOOL;
            case "int" -> FieldType.Primitive.INT;
            case "long" -> FieldType.Primitive.LONG;
            case "float" -> FieldType.Primitive.FLOAT;
            case "string" -> FieldType.Primitive.STRING;
            case "text" -> FieldType.Primitive.TEXT;
            default -> new FieldType.StructRef(type);
        };
    }

    private static void createEmptyCsv(Path dataDir, TableSchema table) throws IOException {
        SchemaToCsvHeader headerGen = new SchemaToCsvHeader();
        headerGen.flattenFields(table.fields());

        List<List<String>> rows = List.of(
                headerGen.getCommentRow(),
                headerGen.getNameRow()
        );

        Path csvPath = dataDir.resolve(table.name() + ".csv");
        CSVUtil.writeToFile(csvPath.toFile(), rows);
    }
}

File: packages/editor-core/src/TableCreateService.ts
/**
 * TableCreateService — TypeScript port of Java `configgen.editorserver.TableCreateService`.
 *
 * Creates new table/struct/enum definitions and writes them back to config.cfg.
 * Optionally creates empty CSV data files for table types.
 *
 * Key differences from Java:
 * - Java methods are static taking (Path dataDir, ...); TS takes an EditorService
 *   instance and uses its rootDir()/context() accessors.
 * - Java uses JSONObject for the request; TS uses a plain object (TableCreateRequest).
 * - Java's SchemaToCsvHeader is not ported; we inline a simple CSV header generator
 *   that handles primitive fields (span=1), which covers all new-table use cases.
 * - TS writes via fs.writeFileSync (synchronous, matching Java's CachedFiles.writeFile).
 *
 * Java source: configgen.editorserver.TableCreateService.java (363 lines)
 */

import * as fs from 'fs';
import * as path from 'path';

import {
  CfgReader,
  CfgWriter,
  CfgSchema,
  CfgSchemaErrs,
  CfgSchemaException,
  ParseError,
  TableSchema,
  StructSchema,
  FieldSchema,
  KeySchema,
  ENo,
  EEnum,
  Primitive,
  StructRef,
  AutoOrPack,
  Metadata_of,
  CommentData,
  metaEnumValuesOfAssigned,
  tableNameNotLowerCase,
  type Nameable,
  type FieldType,
  fieldSpan,
} from '@cfggen/schema';
import { writeCSVToFile, writeCSVToFileAsync, getDefaultFileSystem } from '@cfggen/shared';
import type { EditorService } from './EditorService';

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface CreateResult {
  ok: boolean;
  errors: string[];
}

export interface FieldRequest {
  name: string;
  type?: string;
  comment?: string;
}

export interface EnumValueRequest {
  name: string;
  comment?: string;
}

export interface TableCreateRequest {
  type: 'table' | 'struct' | 'enum';
  name: string;
  fields?: FieldRequest[];
  primaryKey?: string[];
  withDataFile?: boolean;
  enumValues?: EnumValueRequest[];
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

const IDENTIFIER_PATTERN = /^[a-zA-Z_][a-zA-Z0-9_]*$/;

// ---------------------------------------------------------------------------
// TableCreateService
// ---------------------------------------------------------------------------

export class TableCreateService {
  /**
   * Create a new table/struct/enum and write back to config.cfg.
   * Optionally creates an empty CSV data file for table types.
   *
   * Java: TableCreateService.createTable(dataDir, existingText, request)
   */
  static createTable(editor: EditorService, request: TableCreateRequest): CreateResult {
    const errors: string[] = [];

    // 1. Read and parse existing config.cfg from disk
    const cfgPath = path.join(editor.rootDir(), 'config.cfg');
    let existingText = '';
    try {
      if (fs.existsSync(cfgPath)) {
        existingText = fs.readFileSync(cfgPath, 'utf-8');
      }
    } catch (e) {
      errors.push(`Failed to read config.cfg: ${(e as Error).message}`);
      return { ok: false, errors };
    }

    let schema: CfgSchema;
    try {
      if (existingText.trim().length > 0) {
        schema = CfgReader.parse(existingText);
      } else {
        schema = CfgSchema.of();
      }
    } catch (e) {
      if (e instanceof ParseError) {
        errors.push(`Failed to parse existing config.cfg: ${e.message}`);
      } else {
        errors.push(`Failed to parse existing config.cfg: ${(e as Error).message}`);
      }
      return { ok: false, errors };
    }

    // 2. Validate name
    const name = request.name;
    if (!name || name.trim().length === 0) {
      errors.push('Name is required');
      return { ok: false, errors };
    }
    if (!IDENTIFIER_PATTERN.test(name)) {
      errors.push(`Name must be a valid identifier: ${name}`);
      return { ok: false, errors };
    }

    // Check name conflict
    for (const item of schema.items()) {
      if (item.name().toLowerCase() === name.toLowerCase()) {
        errors.push(`Name already exists: ${name}`);
        return { ok: false, errors };
      }
    }

    // 3. Build new element based on type
    let newElement: Nameable | null = null;
    try {
      switch (request.type) {
        case 'table':
          newElement = TableCreateService.buildTableSchema(request, name);
          break;
        case 'struct':
          newElement = TableCreateService.buildStructSchema(request, name);
          break;
        case 'enum':
          newElement = TableCreateService.buildEnumTableSchema(request, name);
          break;
        default:
          errors.push(`Unknown type: ${request.type}, must be 'table', 'struct', or 'enum'`);
          return { ok: false, errors };
      }
    } catch (e) {
      if (e instanceof CfgSchemaException) {
        for (const err of e.errs.errs) {
          errors.push(err.msg());
        }
      } else {
        errors.push((e as Error).message);
      }
      return { ok: false, errors };
    }

    if (newElement === null) {
      return { ok: false, errors };
    }

    // 4. Add to schema and validate
    schema.add(newElement);

    try {
      const errs = schema.resolve();
      errs.checkErrors('createTable');
    } catch (e) {
      if (e instanceof CfgSchemaException) {
        for (const err of e.errs.errs) {
          errors.push(err.msg());
        }
      } else {
        errors.push((e as Error).message);
      }
      return { ok: false, errors };
    }

    // 5. Optionally create empty CSV data file (before writing config.cfg,
    //    so if CSV creation fails, config.cfg is not modified)
    if (request.type === 'table' && request.withDataFile) {
      const table = newElement as TableSchema;
      if (!table.meta().hasEnumValues()) {
        try {
          TableCreateService.createEmptyCsv(editor.rootDir(), table);
        } catch (e) {
          errors.push(`Failed to create data file: ${(e as Error).message}`);
          return { ok: false, errors };
        }
      }
    }

    // 6. Write config.cfg
    try {
      const cfgText = CfgWriter.stringify(schema);
      fs.writeFileSync(cfgPath, cfgText, 'utf8');
    } catch (e) {
      errors.push(`Failed to write config.cfg: ${(e as Error).message}`);
      return { ok: false, errors };
    }

    return { ok: true, errors: [] };
  }

  /**
   * Async variant of createTable.
   * Uses CfgFileSystem abstraction (Tauri/WebView compatible).
   */
  static async createTableAsync(editor: EditorService, request: TableCreateRequest): Promise<CreateResult> {
    const errors: string[] = [];
    const dfs = getDefaultFileSystem();

    // 1. Read and parse existing config.cfg from disk
    const cfgPath = path.join(editor.rootDir(), 'config.cfg');
    let existingText = '';
    try {
      if (await dfs.exists(cfgPath)) {
        const bytes = await dfs.readFile(cfgPath);
        existingText = Buffer.from(bytes).toString('utf-8');
      }
    } catch (e) {
      errors.push(`Failed to read config.cfg: ${(e as Error).message}`);
      return { ok: false, errors };
    }

    let schema: CfgSchema;
    try {
      if (existingText.trim().length > 0) {
        schema = CfgReader.parse(existingText);
      } else {
        schema = CfgSchema.of();
      }
    } catch (e) {
      if (e instanceof ParseError) {
        errors.push(`Failed to parse existing config.cfg: ${e.message}`);
      } else {
        errors.push(`Failed to parse existing config.cfg: ${(e as Error).message}`);
      }
      return { ok: false, errors };
    }

    // 2. Validate name
    const name = request.name;
    if (!name || name.trim().length === 0) {
      errors.push('Name is required');
      return { ok: false, errors };
    }
    if (!IDENTIFIER_PATTERN.test(name)) {
      errors.push(`Name must be a valid identifier: ${name}`);
      return { ok: false, errors };
    }

    // Check name conflict
    for (const item of schema.items()) {
      if (item.name().toLowerCase() === name.toLowerCase()) {
        errors.push(`Name already exists: ${name}`);
        return { ok: false, errors };
      }
    }

    // 3. Build new element based on type
    let newElement: Nameable | null = null;
    try {
      switch (request.type) {
        case 'table':
          newElement = TableCreateService.buildTableSchema(request, name);
          break;
        case 'struct':
          newElement = TableCreateService.buildStructSchema(request, name);
          break;
        case 'enum':
          newElement = TableCreateService.buildEnumTableSchema(request, name);
          break;
        default:
          errors.push(`Unknown type: ${request.type}, must be 'table', 'struct', or 'enum'`);
          return { ok: false, errors };
      }
    } catch (e) {
      if (e instanceof CfgSchemaException) {
        for (const err of e.errs.errs) {
          errors.push(err.msg());
        }
      } else {
        errors.push((e as Error).message);
      }
      return { ok: false, errors };
    }

    if (newElement === null) {
      return { ok: false, errors };
    }

    // 4. Add to schema and validate
    schema.add(newElement);

    try {
      const errs = schema.resolve();
      errs.checkErrors('createTable');
    } catch (e) {
      if (e instanceof CfgSchemaException) {
        for (const err of e.errs.errs) {
          errors.push(err.msg());
        }
      } else {
        errors.push((e as Error).message);
      }
      return { ok: false, errors };
    }

    // 5. Optionally create empty CSV data file
    if (request.type === 'table' && request.withDataFile) {
      const table = newElement as TableSchema;
      if (!table.meta().hasEnumValues()) {
        try {
          await TableCreateService.createEmptyCsvAsync(editor.rootDir(), table);
        } catch (e) {
          errors.push(`Failed to create data file: ${(e as Error).message}`);
          return { ok: false, errors };
        }
      }
    }

    // 6. Write config.cfg
    try {
      const cfgText = CfgWriter.stringify(schema);
      await dfs.writeFile(cfgPath, Buffer.from(cfgText, 'utf8'));
    } catch (e) {
      errors.push(`Failed to write config.cfg: ${(e as Error).message}`);
      return { ok: false, errors };
    }

    return { ok: true, errors: [] };
  }

  /**
   * Java: TableCreateService.createDataFile(dataDir, tableName)
   */
  static createDataFile(editor: EditorService, tableName: string): CreateResult {
    const errors: string[] = [];

    // Read and parse config.cfg
    const cfgPath = path.join(editor.rootDir(), 'config.cfg');
    let cfgText: string;
    try {
      cfgText = fs.readFileSync(cfgPath, 'utf-8');
    } catch (e) {
      errors.push(`Failed to read config.cfg: ${(e as Error).message}`);
      return { ok: false, errors };
    }

    let schema: CfgSchema;
    try {
      schema = CfgReader.parse(cfgText);
      schema.resolve().checkErrors('createDataFile');
    } catch (e) {
      errors.push(`Failed to parse/resolve schema: ${(e as Error).message}`);
      return { ok: false, errors };
    }

    // Find target table
    let targetTable: TableSchema | null = null;
    for (const item of schema.items()) {
      if (item instanceof TableSchema && item.name() === tableName) {
        targetTable = item;
        break;
      }
    }

    if (targetTable === null) {
      errors.push(`Table not found: ${tableName}`);
      return { ok: false, errors };
    }

    if (targetTable.isJson()) {
      errors.push(`Table uses JSON data source, cannot create CSV file: ${tableName}`);
      return { ok: false, errors };
    }

    if (targetTable.meta().hasEnumValues()) {
      errors.push(`Enum table does not need a data file: ${tableName}`);
      return { ok: false, errors };
    }

    // Check if file already exists
    const csvPath = path.join(editor.rootDir(), tableName + '.csv');
    if (fs.existsSync(csvPath)) {
      errors.push(`Data file already exists: ${path.basename(csvPath)}`);
      return { ok: false, errors };
    }

    try {
      TableCreateService.createEmptyCsv(editor.rootDir(), targetTable);
    } catch (e) {
      errors.push(`Failed to create data file: ${(e as Error).message}`);
      return { ok: false, errors };
    }

    return { ok: true, errors: [] };
  }

  /**
   * Async variant of createDataFile.
   * Uses CfgFileSystem abstraction (Tauri/WebView compatible).
   */
  static async createDataFileAsync(editor: EditorService, tableName: string): Promise<CreateResult> {
    const errors: string[] = [];
    const dfs = getDefaultFileSystem();

    // Read and parse config.cfg
    const cfgPath = path.join(editor.rootDir(), 'config.cfg');
    let cfgText: string;
    try {
      const bytes = await dfs.readFile(cfgPath);
      cfgText = Buffer.from(bytes).toString('utf-8');
    } catch (e) {
      errors.push(`Failed to read config.cfg: ${(e as Error).message}`);
      return { ok: false, errors };
    }

    let schema: CfgSchema;
    try {
      schema = CfgReader.parse(cfgText);
      schema.resolve().checkErrors('createDataFile');
    } catch (e) {
      errors.push(`Failed to parse/resolve schema: ${(e as Error).message}`);
      return { ok: false, errors };
    }

    // Find target table
    let targetTable: TableSchema | null = null;
    for (const item of schema.items()) {
      if (item instanceof TableSchema && item.name() === tableName) {
        targetTable = item;
        break;
      }
    }

    if (targetTable === null) {
      errors.push(`Table not found: ${tableName}`);
      return { ok: false, errors };
    }

    if (targetTable.isJson()) {
      errors.push(`Table uses JSON data source, cannot create CSV file: ${tableName}`);
      return { ok: false, errors };
    }

    if (targetTable.meta().hasEnumValues()) {
      errors.push(`Enum table does not need a data file: ${tableName}`);
      return { ok: false, errors };
    }

    // Check if file already exists
    const csvPath = path.join(editor.rootDir(), tableName + '.csv');
    if (await dfs.exists(csvPath)) {
      errors.push(`Data file already exists: ${path.basename(csvPath)}`);
      return { ok: false, errors };
    }

    try {
      await TableCreateService.createEmptyCsvAsync(editor.rootDir(), targetTable);
    } catch (e) {
      errors.push(`Failed to create data file: ${(e as Error).message}`);
      return { ok: false, errors };
    }

    return { ok: true, errors: [] };
  }
  // -------------------------------------------------------------------------

  private static buildTableSchema(request: TableCreateRequest, name: string): TableSchema {
    // Table name must be lowercase
    if (name !== name.toLowerCase()) {
      const errs = CfgSchemaErrs.of();
      errs.addErr(tableNameNotLowerCase(name));
      throw new CfgSchemaException(errs);
    }

    const fields = TableCreateService.buildFields(request.fields);
    if (fields.length === 0) {
      throw new Error('Table must have at least one field');
    }

    // Primary key
    let primaryKey: KeySchema;
    if (request.primaryKey && request.primaryKey.length > 0) {
      primaryKey = new KeySchema(request.primaryKey);
    } else {
      primaryKey = new KeySchema([fields[0].name]);
    }

    return new TableSchema(
      name.toLowerCase(),
      primaryKey,
      ENo.NO,
      false,
      Metadata_of(),
      fields,
      [],
      [],
    );
  }

  private static buildStructSchema(request: TableCreateRequest, name: string): StructSchema {
    const fields = TableCreateService.buildFields(request.fields);
    if (fields.length === 0) {
      throw new Error('Struct must have at least one field');
    }

    return new StructSchema(
      name,
      AutoOrPack.AUTO,
      Metadata_of(),
      fields,
      [],
    );
  }

  private static buildEnumTableSchema(request: TableCreateRequest, name: string): TableSchema {
    // Table name must be lowercase
    if (name !== name.toLowerCase()) {
      const errs = CfgSchemaErrs.of();
      errs.addErr(tableNameNotLowerCase(name));
      throw new CfgSchemaException(errs);
    }

    const enumValues = request.enumValues;
    if (!enumValues || enumValues.length === 0) {
      throw new Error('Enum must have at least one value');
    }

    const values = enumValues.map((v, i) => ({
      name: v.name,
      comment: v.comment ?? '',
      number: i,
    }));

    const meta = Metadata_of();
    meta.putEnumValues(metaEnumValuesOfAssigned(values));

    // Enum's field is "name"
    const nameField = new FieldSchema(
      'name',
      Primitive.STRING,
      AutoOrPack.AUTO,
      Metadata_of(),
    );

    return new TableSchema(
      name.toLowerCase(),
      new KeySchema(['name']),
      new EEnum('name'),
      false,
      meta,
      [nameField],
      [],
      [],
    );
  }

  private static buildFields(fieldRequests?: FieldRequest[]): FieldSchema[] {
    if (!fieldRequests) {
      return [];
    }

    const fields: FieldSchema[] = [];
    for (const f of fieldRequests) {
      const fieldName = f.name;
      if (!fieldName || fieldName.trim().length === 0) {
        throw new Error('Field name is required');
      }
      if (!IDENTIFIER_PATTERN.test(fieldName)) {
        throw new Error(`Invalid field name: ${fieldName}`);
      }

      const fieldType = TableCreateService.parseFieldType(f.type ?? 'int');
      const meta = Metadata_of();

      if (f.comment && f.comment.trim().length > 0) {
        meta.putComment(new CommentData('', f.comment, null));
      }

      fields.push(new FieldSchema(fieldName, fieldType, AutoOrPack.AUTO, meta));
    }
    return fields;
  }

  private static parseFieldType(type: string): FieldType {
    switch (type.toLowerCase()) {
      case 'bool':
        return Primitive.BOOL;
      case 'int':
        return Primitive.INT;
      case 'long':
        return Primitive.LONG;
      case 'float':
        return Primitive.FLOAT;
      case 'string':
      case 'str':
        return Primitive.STRING;
      case 'text':
        return Primitive.TEXT;
      default:
        return new StructRef(type);
    }
  }

  // -------------------------------------------------------------------------
  // Internal: CSV creation
  // -------------------------------------------------------------------------

  /**
   * Create an empty CSV file with header rows (comment row + name row).
   * Simplified version of Java's SchemaToCsvHeader — only handles flat fields
   * (span=1), which is sufficient for newly created tables.
   */
  private static createEmptyCsv(rootDir: string, table: TableSchema): void {
    const commentRow: string[] = [];
    const nameRow: string[] = [];

    for (const field of table.fields()) {
      const span = fieldSpan(field);
      const comment = field.comment() ?? '';

      if (span === 1) {
        // Single column field
        nameRow.push(field.name);
        commentRow.push(comment);
      } else {
        // Multi-span field (list, map, struct ref) — use _-prefixed names
        // For simple new tables this path won't be hit, but handle gracefully
        for (let i = 0; i < span; i++) {
          nameRow.push(i === 0 ? field.name : `_${i}`);
          commentRow.push(i === 0 ? comment : '');
        }
      }
    }

    // If no fields (shouldn't happen for valid tables), add at least one column
    if (nameRow.length === 0) {
      nameRow.push('id');
      commentRow.push('');
    }

    const csvPath = path.join(rootDir, table.name() + '.csv');
    writeCSVToFile(csvPath, [commentRow, nameRow]);
  }

  /**
   * Async variant of createEmptyCsv.
   * Uses CfgFileSystem abstraction (Tauri/WebView compatible).
   */
  private static async createEmptyCsvAsync(rootDir: string, table: TableSchema): Promise<void> {
    const commentRow: string[] = [];
    const nameRow: string[] = [];

    for (const field of table.fields()) {
      const span = fieldSpan(field);
      const comment = field.comment() ?? '';

      if (span === 1) {
        nameRow.push(field.name);
        commentRow.push(comment);
      } else {
        for (let i = 0; i < span; i++) {
          nameRow.push(i === 0 ? field.name : `_${i}`);
          commentRow.push(i === 0 ? comment : '');
        }
      }
    }

    if (nameRow.length === 0) {
      nameRow.push('id');
      commentRow.push('');
    }

    const csvPath = path.join(rootDir, table.name() + '.csv');
    await writeCSVToFileAsync(csvPath, [commentRow, nameRow]);
  }
}
